### ISSUE TITLE
Annotations not displaying accurate time over multiple days view

### ISSUE THREAD

**Issue by alexganwd** [grafana/grafana#15764]
### What Grafana version are you using?
6.0

### What datasource are you using?
Graphite

### What OS are you running grafana on?
Ubuntu 14.04

### What did you do?
Used annotations API to show deployments in dashboards using regions. 

### What was the expected result?
Region annotation showing up in the graph matching the deployment time.
### What happened instead?
Every often, we see some of these annotations which are meant to show regions no bigger than 20 minutes covering the full graph(7 days), this only happen when the date criteria is selected is to more than 24h . Below image is an example of this:

<img width="1595" alt="annotation_using_full_graph" src="https://user-images.githubusercontent.com/10225769/53741220-d10b4b80-3e8d-11e9-9c9c-c198a98bf438.png">

However if I zoom in: 
<img width="1010" alt="screenshot 2019-03-04 at 14 49 16" src="https://user-images.githubusercontent.com/10225769/53741284-f13b0a80-3e8d-11e9-875b-803467f98752.png">

This is the annotation, I get when I query the annotations API:

    {
        "id": 22038,
        "alertId": 0,
        "alertName": "",
        "dashboardId": 0,
        "panelId": 0,
        "userId": 0,
        "newState": "",
        "prevState": "",
        "created": 1551697818876,
        "updated": 1551697818876,
        "time": 1551697819950,
        "text": "Deploying {'myservice-1': 'xxx', 'myservice-2': 'yyyy', 'myservice-3': 'ccc'}",
        "regionId": 22037,
        "tags": [
            "myservice",
            "deployment",
            "myenv"
        ],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {}
    },
    {
        "id": 22037,
        "alertId": 0,
        "alertName": "",
        "dashboardId": 0,
        "panelId": 0,
        "userId": 0,
        "newState": "",
        "prevState": "",
        "created": 1551697818843,
        "updated": 1551697818858,
        "time": 1551697624709,
        "text": "Deploying {'myservice-1': 'xxx', 'myservice-2': 'yyyy', 'myservice-3': 'ccc'}",
        "regionId": 22037,
        "tags": [
            "myservice",
            "deployment",
            "myenv"
        ],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {}
    },

**Comment by marefr** [comment]
@alexganwd do you still experience this issue? Unable to reproduce this. Maybe it's related to that you have multiple annotations. Do you have any specific setting for the graph panel like relative time? If still a problem, could you include the url and the response of annotations request used to render a broken graph as described above (similar to `http://localhost:3000/api/annotations?from=612397993014&limit=100&matchAny=true&tags=deploy&to=2499834793014`) - you can use chrome developer tools/network tab to find the url/response. Thanks

**Comment by bradbyu** [comment]
@marefr We're using 5.4.x currently and experiencing the same problem. Attached is
[grafana.txt](https://github.com/grafana/grafana/files/3112962/grafana.txt)
 one the the responses where it shows for us. There was also an angular.js error
`angular.js:14700 Error: [ngRepeat:dupes] Duplicates in a repeater are not allowed. Use 'track by' expression to specify unique keys. Repeater: tag in tags, Duplicate key: string:ok, Duplicate value: ok
http://errors.angularjs.org/1.6.6/ngRepeat/dupes?p0=tag%20in%20tags&p1=string%3Aok&p2=ok
    at angular.js:116
    at angular.js:31449
    at angular.js:18121
    at l.$digest (angular.js:18260)
    at jquery.flot.events.ts:165
    at Object.invoke (angular.js:5040)
    at Xi (jquery.flot.events.ts:154)
    at HTMLDivElement.<anonymous> (graph_tooltip.ts:21)
    at HTMLDivElement.handle (jquery.js:5536)
    at HTMLDivElement.dispatch (jquery.js:5183)`

**Comment by marefr** [comment]
@bradbyu thanks, but you've included the response of the metric query - was asking for the response of the annotations query.

**Comment by bradbyu** [comment]
@marefr Sorry about that. I think this is the correct response you're looking for (for annotations?from=1556109380607&limit=300&tags=cronjob&to=1556120180607). Let me know if not.
[grafana-response.txt](https://github.com/grafana/grafana/files/3113088/grafana-response.txt)

**Comment by marefr** [comment]
@bradbyu thanks. And which one of those are the problematic one? Can you include screenshot?

**Comment by bradbyu** [comment]
@marefr It never seems to be the same one. These are annotations sent from recurring processes. Sometimes the whole graph shows red, sometimes it doesn't. Here's a screenshot from that same response window with me hovering over the "long tail"
![Screen Shot 2019-04-24 at 12 52 46 PM](https://user-images.githubusercontent.com/5098494/56678310-5d690c00-6690-11e9-8abc-01ea4157e40f.png)
If I zoom in to annotations?from=1556109380607&limit=300&tags=cronjob&to=1556119577928 the problem doesn't show.
[grafana-response2.txt](https://github.com/grafana/grafana/files/3113396/grafana-response2.txt)
![Screen Shot 2019-04-24 at 1 03 21 PM](https://user-images.githubusercontent.com/5098494/56678717-5e4e6d80-6691-11e9-9163-0a68a7e8e349.png)
If I zoom in further to annotations?from=1556109380607&limit=300&tags=cronjob&to=1556118557000 the problem shows again, but with a different "long tail" on hover.
[grafana-response3.txt](https://github.com/grafana/grafana/files/3113408/grafana-response3.txt)
![Screen Shot 2019-04-24 at 1 06 47 PM](https://user-images.githubusercontent.com/5098494/56678991-f8161a80-6691-11e9-820b-7a5cf1e4beba.png)

**Comment by alexganwd** [comment]
@marefr  We are still experiencing the same issue. However, I have just realised that this is only happening in a dashboard where the interval time (time range) is set to 7 days but some graphs are using a relative time of 1d.

**Comment by marefr** [comment]
@bradbyu thanks for the responses. Was able to use your first response as a mock in the code instead of actual annotation response. I get the following graph result (similar to your's but different times probably due to timezone):

![image](https://user-images.githubusercontent.com/1668778/56732090-8e167780-675c-11e9-8a29-b2a9c23b44ba.png)

The region spanning from the beginning of the graph is originating from this data:
```json
{
  "id": 53850,
  "alertId": 0,
  "alertName": "",
  "dashboardId": 0,
  "panelId": 0,
  "userId": 0,
  "newState": "",
  "prevState": "",
  "created": 1556117821495,
  "updated": 1556117830595,
  "time": 1556117830284,
  "text": "add items [reverb]",
  "regionId": 53849,
  "tags": ["cronjob", "web-batch01", "ok"],
  "login": "",
  "email": "",
  "avatarUrl": "",
  "data": {}
}
```

This data only have one item where all the other region annotations have two. So it seems to me that the region, with regionId=53849, has a `time` before 1556109380607. According to our source code this is  expected behavior:

https://github.com/grafana/grafana/blob/25ea5fcaf2dc7dbf341d7c905b0a31dbd8277cbd/public/app/features/annotations/specs/annotations_srv_specs.test.ts#L31-L38

@bradbyu to me it seems like your reported problem is not a bug, but if you think it still is you'll need to find the corresponding annotation item with regionId=53849.

@alexganwd if you can provide annotation response of your problem I can verify those as well.

**Comment by bradbyu** [comment]
@marefr I think the issue is with the tags limit. In the previously linked files grafana-response2.txt and grafana-response3.txt (which is a shorter time window) I found the regionId=53849
`    {
        "id": 53849,
        "alertId": 0,
        "alertName": "",
        "dashboardId": 0,
        "panelId": 0,
        "userId": 0,
        "newState": "",
        "prevState": "",
        "created": 1556117821483,
        "updated": 1556117830386,
        "time": 1556117821319,
        "text": "add items [reverb]",
        "regionId": 53849,
        "tags": [
            "cronjob",
            "web-batch01",
            "ok"
        ],
        "login": "",
        "email": "",
        "avatarUrl": "",
        "data": {}
    }`

**Comment by marefr** [comment]
@bradbyu thanks. This is weird, but don't exactly understand what you mean by tags limit - it's the same for all three requests. Do you mean that the limit of 300 is to small? That's the only explanation I have since have looked at the database query and that looks correct.

**Comment by bradbyu** [comment]
@marefr The limit is working correctly (bringing back exactly 300 json objects), BUT because of that, the display seems to assume that if there isn't an _id_ to match a referenced _regionId_, then it assumes the matching _regionId_ is "out-of-bounds" (so to speak) and displays as if it took place before the _from_ query parameter. Those bounds (_from_ and _to_) always contain the "missing" _regionId_ (at least in our case), but because the query is only returning the _limit_, it acts as if it is missing.

So in the query, I think the solution here is:

- Return up to the limit provided (i.e. 300)

- Match up the json objects with regions

- If there are json objects with regions that don't match with anything within the limit, then either

1. Find the matching json object(s) to the region and include it in the results (i.e. thus returning more than 300 json objects)

1. Throw the unmatched json object(s) out of the results (i.e. thus returning less than 300 json objects)

**Comment by alexganwd** [comment]
@marefr ,  please find attached my annotations response. Please let me know if there is anything else I can add. 
The problem on this graph is with annotation: id: 44601. It's showing over a period of 7h
[annotations_response.txt](https://github.com/grafana/grafana/files/3126896/annotations_response.txt)

**Comment by bradbyu** [comment]
@marefr I have a few more things to note

1. We are planning on using annotations more, so there will likely be more than 1 mismatched annotation in the limit result set, so it is important to handle multiple mismatches.

1. We just updated to the latest 6.x Grafana and the issue is still present.

**Comment by marefr** [comment]
Now I understand the problem. A region annotation uses two rows in database table where's a non-region annotation used one row. Because how the annotation query is implemented right now with limit of 300 there's a chance that non-complete (2 rows) region annotations are returned from the database query/API and this is causing your reported problem.

I can see that this feels like a bug, but given when and how it was implemented using a limit of 300 set which at that time seemed like a very high value I will label this as a feature request. 

Given the current roadmap this is not something that we are planning to work with. If anyone is interested in trying to solve this the code in question is:
https://github.com/grafana/grafana/blob/e5971eef1f92b40a78457f6fa17a3878533ef723/pkg/services/sqlstore/annotation.go#L118-L236

**Comment by ryantxu** [comment]
@alexganwd -- I think this is fixed for 6.4, please reopen if not.  See #17673
